v2/controllers/pod_watcher.go

```diff
@@ -178,23 +193,61 @@ OUTER2:
 	}
 
 	r.podAddrs[key] = podIPs
+	for _, ip := range podIPs {
+		keySet, ok := r.peers[ip.String()]
+		if !ok {
+			keySet = map[string]struct{}{}
+			r.peers[ip.String()] = keySet
+		}
+		keySet[key] = struct{}{}
+	}
+
 	r.metric.Set(float64(len(r.podAddrs)))
 	return nil
 }
 
+func (r *podWatcher) delTerminatedPod(pod *corev1.Pod, logger logr.Logger) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	return r.delPeer(fmt.Sprintf("%s/%s", pod.Namespace, pod.Name), "delTerminatedPod", string(pod.Status.Phase), logger)
+}
+
 func (r *podWatcher) delPod(n types.NamespacedName, logger logr.Logger) error {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 
-	key := n.Namespace + "/" + n.Name
+	return r.delPeer(fmt.Sprintf("%s/%s", n.Namespace, n.Name), "delPod", "", logger)
+}
+
+func (r *podWatcher) delPeer(key, caller, podStatus string, logger logr.Logger) error {
 	for _, ip := range r.podAddrs[key] {
-		logger.Info("delete peer", "caller", "delPod", "ip", ip.String(), "podIPs", r.podAddrs[key])
-		if err := r.ft.DelPeer(ip); err != nil {
-			return err
+		if !r.existsOtherLivePeers(key, ip.String()) {
+			logger.Info("delete peer", "caller", caller, "ip", ip.String(), "podIPs", r.podAddrs[key], "podPhase", podStatus)
+			if err := r.ft.DelPeer(ip); err != nil {
+				return err
+			}
+		}
+
+		if keySet, ok := r.peers[ip.String()]; ok {
+			delete(keySet, key)
+			if len(keySet) == 0 {
+				delete(r.peers, ip.String())
+			}
 		}
 	}
 
 	delete(r.podAddrs, key)
 	r.metric.Set(float64(len(r.podAddrs)))
 	return nil
 }
+
+func (r *podWatcher) existsOtherLivePeers(key, ip string) bool {
+	if keySet, ok := r.peers[ip]; ok {
+		if _, ok := keySet[key]; ok {
```

Comment:
I suspect `r.peers` always contains `ip`, and `keySet` always contains `key`.
Don't we need to return an error in other situations? Or do different scenarios exist?

Reply:
There's no situation I can think of in which `r.peers` doesn't have `ip` nor `keySet` doesn't have `key` even though `r.podAddrs` has the key. If it happened, it means something wrong occurred and broke the integrity. So I think we should abort deleting a peer in that unexpected situation. Is that what you are concerned about?